labs/src/main/scala/org/scalalabs/basic/lab02/ListManipulationExercise02.scala

```diff
@@ -38,30 +36,12 @@ object ListManipulationExercise02 {
    * may be able to achieve the same functionality as implemented below
    * in a one-liner.
    */
-  def separateTheYoungFromTheOld(persons: List[Person]): List[List[String]] = {
-    var youngins: ListBuffer[Person] = new ListBuffer[Person]()
-    var elders: ListBuffer[Person] = new ListBuffer[Person]()
-    var validYoungNames: ListBuffer[String] = new ListBuffer[String]()
-    var validOldNames: ListBuffer[String] = new ListBuffer[String]()
-
-    for (person <- persons) {
-        if (person.age < 18) {
-          youngins += person
-        } else {
-          elders += person
-        }
-    }
-
-    var sortedYoung = youngins.toList.sortBy(_.age)
-    var sortedOld = elders.toList.sortBy(_.age)
-
-    for (young <- sortedYoung) {
-      validYoungNames += young.firstName
-    }
-    for (old <- sortedOld) {
-      validOldNames += old.firstName
-    }
-    List(validYoungNames.toList, validOldNames.toList)
-  }
+  def separateTheYoungFromTheOld(persons: List[Person]): List[List[String]] = 
+  	if (persons.isEmpty)
+  	  List(List(), List())
+    else if (persons.head.age < 18) 
+      List((persons.head.firstName :: separateTheYoungFromTheOld(persons.tail).head).sorted, (separateTheYoungFromTheOld(persons.tail).last).sorted)
```

Comment:
With either recursion or higher-order functions, it’s more efficient if you don’t recompute things. In the recursive case you can avoid that by saving the result of `separateTheYoungFromTheOld(persons.tail)` in a `val`, e.g. `val List(young, old) = separateTheYoungFromTheOld(persons.tail)`.